HPXML-based workflow for residential buildings

joseph-robertson opened the pull request:
Summary of changes:
* Fields added:
  * `foundation_type`
    * slab
    * crawlspace - vented
    * crawlspace - unvented
    * basement - unconditioned
    * basement - conditioned
    * ambient
  * `attic_type`
    * attic - vented
    * attic - unvented
    * attic - conditioned
  * `number_of_bedrooms`
    * integer
  * `heating_system_fuel_type`
    * electricity
    * natural gas
    * fuel oil
    * propane
    * wood
* Fields updated:
  * `roof_type`
    * Changed from free entry field to enums: gable, hip, flat
  * `system_type`
    * Removed existing "Residential" enums
    * Added a bunch of "Residential - ..." heating/cooling system combination enums
* Other schema updates:
  * "Single-Family" changed to "Single-Family Detached"
  * "Multifamily (2 to 4 units)" changed to "Single-Family Attached"
  * "Multifamily (5 or more units)" changed to "Multifamily"
  * Add a "oneOf" element to schema for enforcing "requires" by `building_type`:
    * Single-Family Detached
      * id, type, name, floor_area, number_of_stories, footprint_area, building_type, number_of_bedrooms, foundation_type, attic_type, roof_type, heating_system_fuel_type
    * Single-Family Attached, Multifamily
      * id, type, name, floor_area, number_of_stories, footprint_area, building_type, number_of_residential_units, number_of_bedrooms, heating_system_fuel_type
    * Commercial building types
      * id, type, name, floor_area, number_of_stories, footprint_area, building_type (these are unchanged from before)

shorowit commented on lib/urbanopt/geojson/schema/building_properties.json:
```diff
@@ -332,15 +415,32 @@
+        "Residential - no heating and evaporative cooler",
+        "Residential - air-to-air heat pump",
+        "Residential - mini-split heat pump",
+        "Residential - ground-to-air heat pump"
```
I would drop any choices without heating systems, that is exceedingly rare. Perhaps offer electric resistance instead.

shorowit commented on lib/urbanopt/geojson/schema/building_properties.json:
```diff
@@ -332,15 +415,32 @@
+      ]
+    },
+    "heatingSystemFuelType": {
+      "description": "The fuel type of the heating system.",
```
Add a note that this may not apply for certain system types (e.g., heat pumps or electric resistance)?

shorowit commented on lib/urbanopt/geojson/schema/building_properties.json:
```diff
@@ -376,6 +476,36 @@
+      ]
+    },
+    "atticType": {
+      "description": "The attic type of the building. Required for residential buildings.",
```
Attic type won't apply if there's a flat roof. And choice between gable and hip isn't really so important (other than geometry rendering). I suggest combining attic/roof choices into a single list:
- "attic - vented"
- "attic - unvented"
- "attic - conditioned"
- "flat roof"

And then just assume home has a gable roof if there's an attic.

joseph-robertson commented:
UPDATE:

Summary of changes:
* Fields added:
  * `foundation_type`
    * slab
    * crawlspace - vented
    * crawlspace - unvented
    * basement - unconditioned
    * basement - conditioned
    * ambient
  * `attic_type`
    * attic - vented
    * attic - unvented
    * attic - conditioned
    * flat roof
  * `number_of_bedrooms`
    * integer
  * `heating_system_fuel_type`
    * electricity
    * natural gas
    * fuel oil
    * propane
    * wood
* Fields updated:
  * `system_type`
    * Removed existing "Residential" enums
    * Added a bunch of "Residential - ..." heating/cooling system combination enums
* Other schema updates:
  * "Single-Family" changed to "Single-Family Detached"
  * "Multifamily (2 to 4 units)" changed to "Single-Family Attached"
  * "Multifamily (5 or more units)" changed to "Multifamily"
  * Add a "oneOf" element to schema for enforcing "requires" by `building_type`:
    * Single-Family Detached
      * id, type, name, floor_area, number_of_stories, footprint_area, building_type, foundation_type, attic_type
    * Single-Family Attached
      * id, type, name, floor_area, number_of_stories, footprint_area, building_type, foundation_type, attic_type, number_of_residential_units, number_of_bedrooms
    * Multifamily
      * id, type, name, floor_area, number_of_stories, footprint_area, building_type, foundation_type, number_of_residential_units, number_of_bedrooms
    * Commercial building types
      * id, type, name, floor_area, number_of_stories, footprint_area, building_type (these are unchanged from before)

shorowit commented on lib/urbanopt/geojson/schema/building_properties.json:
```diff
@@ -185,7 +194,11 @@
+      "type": "integer"
+    },
+    "number_of_bedrooms": {
+      "description": "Total number of bedrooms in the building. Required for single-family attached or multifamily residential buildings. Must be divisible by the number of residential units.",
```
I don't understand, why is SFD excluded?

kflemin commented on lib/urbanopt/geojson/schema/building_properties.json:
```diff
@@ -388,6 +515,28 @@
     }
   }
-}
+}
```
add a hard-return at the end of this file

kflemin commented:
Looks good! after you add that hard-return at the end of the building_properties.json file, you can merge and delete this branch. thanks!
